src/java/org/apache/cassandra/cql3/Marker.java

```diff
@@ -0,0 +1,128 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.cql3;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import org.apache.cassandra.cql3.functions.Function;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.MultiElementType;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+import static org.apache.cassandra.cql3.statements.RequestValidations.checkFalse;
+
+/**
+ * A placeholder, also called bind marker, for a single value represented in CQL as '?' for an unnamed marker or ':<name>' for a named marker.
+ * For example, 'SELECT ... WHERE pk = ?' or 'SELECT ... WHERE pk = :myKey '.
+ */
+public final class Marker extends Term.NonTerminal
+{
+    private final int bindIndex;
+
+    private final ColumnSpecification receiver;
+
+    private Marker(int bindIndex, ColumnSpecification receiver)
+    {
+        this.bindIndex = bindIndex;
+        this.receiver = receiver;
+    }
+
+    @Override
+    public void collectMarkerSpecification(VariableSpecifications boundNames)
+    {
+        boundNames.add(bindIndex, receiver);
+    }
+
+    @Override
+    public boolean containsBindMarker()
+    {
+        return true;
+    }
+
+    @Override
+    public void addFunctionsTo(List<Function> functions)
+    {
+    }
+
+    @Override
+    public Term.Terminal bind(QueryOptions options) throws InvalidRequestException
+    {
+        try
+        {
+            ByteBuffer bytes = options.getValues().get(bindIndex);
+            if (bytes == null)
+                return null;
+
+            if (bytes == ByteBufferUtil.UNSET_BYTE_BUFFER)
+            {
+                checkFalse(receiver.type.isTuple(), "Invalid unset value for tuple %s", receiver.name);
+                return Constants.UNSET_VALUE;
+            }
+
+            if (receiver.type instanceof MultiElementType<?>)
+                return MultiElements.Value.fromSerialized(bytes, (MultiElementType<?>) receiver.type);
+
+            receiver.type.validate(bytes);
+            return new Constants.Value(bytes);
+        }
+            catch (MarshalException e)
+        {
+            throw new InvalidRequestException(e.getMessage(), e);
```

Comment:
`RequestValidations` provides utility methods to perform validation checks of requests. Part of those methods is the formatting of message which is why they usually take as arguments `String messageTemplate, Object... messageArgs`. So rather than using :` new InvalidRequestException(String.format("You should not use %s in a CQL query", x));` you can use `invalidRequest("You should not use %s in a CQL query", x)`. In this case it is just a conversion of a MarshalException into InvalidRequestException so it is not a method that `RequestValidations` provides and I am not sure of the added value to create a method for that.